pkg/operators/clustermanager/controllers/crdstatuccontroller/crd_status_controller.go

```diff
@@ -0,0 +1,173 @@
+	}
+)
+
+type crdStatusController struct {
+	kubeconfig                *rest.Config
+	kubeClient                kubernetes.Interface
+	clusterManagerLister      operatorlister.ClusterManagerLister
+	generateHubClusterClients func(hubConfig *rest.Config) (apiextensionsclient.Interface, error)
+}
+
+// NewClusterManagerController construct cluster manager hub controller
+func NewCRDStatusController(
+	kubeconfig *rest.Config,
+	kubeClient kubernetes.Interface,
+	clusterManagerInformer operatorinformer.ClusterManagerInformer,
+	recorder events.Recorder) factory.Controller {
+	controller := &crdStatusController{
+		kubeconfig:                kubeconfig,
+		kubeClient:                kubeClient,
+		clusterManagerLister:      clusterManagerInformer.Lister(),
+		generateHubClusterClients: generateHubClients,
+	}
+
+	return factory.New().WithSync(controller.sync).
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterManagerInformer.Informer()).
+		ToController("CRDStatusController", recorder)
+}
+
+func (c *crdStatusController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	clusterManagerName := controllerContext.QueueKey()
+	klog.V(4).Infof("Reconciling ClusterManager %q", clusterManagerName)
+
+	clusterManager, err := c.clusterManagerLister.Get(clusterManagerName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// ClusterManager is deleting
+	if !clusterManager.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// need to wait storage version migrations succeed.
+	if succeeded := meta.IsStatusConditionTrue(clusterManager.Status.Conditions, migrationcontroller.MigrationSucceeded); !succeeded {
+		controllerContext.Queue().AddAfter(clusterManagerName, 5*time.Second)
+		klog.V(4).Info("Wait storage version migration succeed.")
+		return nil
+	}
+
+	// If mode is default, then config is management kubeconfig, else it would use management kubeconfig to find the hub
+	hubKubeconfig, err := helpers.GetHubKubeconfig(ctx, c.kubeconfig, c.kubeClient, clusterManager.Name, clusterManager.Spec.DeployOption.Mode)
+	if err != nil {
+		return err
+	}
+
+	apiExtensionClient, err := c.generateHubClusterClients(hubKubeconfig)
+	if err != nil {
+		return err
+	}
+
+	err = updateStoredVersion(ctx, apiExtensionClient)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateStoredVersion update(remove) deleted api version from CRD status.StoredVersions
+func updateStoredVersion(ctx context.Context, hubAPIExtensionClient apiextensionsclient.Interface) error {
+	for name, desiredVersions := range desiredCRDStoredVersions {
+		// retrieve CRD
+		crd, err := hubAPIExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		isInclude := includeStrings(crd.Status.StoredVersions, desiredVersions)
```

Comment:
I think we just need deepequal?

Reply:
Changed.
As the `StoredVersions` is a `[]string`, and deepequal can not  used here directly. 
So I use a `sets.string` to convert the `[]string`,  then use deepequal to check it.

Reply:
Curious: why we cannot use reflect.DeepEqual against []string?

Reply:
> Curious: why we cannot use reflect.DeepEqual against []string?

For example:
```
s1[0]="v1beta1", s1[1]="v1beta2"
s2[0]="v1beta2", s2[1]="v1beta1"
```

In our Scenario, s1 should be same as s2. but reflect.DeepEqual(s1,s2)  will return false.

Reply:
How does string array be disordered? I think kube crd controller will update storedversion with an ordered string array.

Reply:
You are right. 
Changed to use reflect.DeepEqual to compare the array directly.